rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbDeviceStateNode.java

```diff
@@ -0,0 +1,197 @@
+                "<br>" +
+                "Supported device connectivity events are:" +
+                "<ul>" +
+                "<li>Connect event</li>" +
+                "<li>Disconnect event</li>" +
+                "<li>Activity event</li>" +
+                "<li>Inactivity event</li>" +
+                "</ul>" +
+                "This node is particularly useful when device isn't using transports to receive data, such as when fetching data from external API or computing new data within the rule chain.",
+        configClazz = TbDeviceStateNodeConfiguration.class,
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeviceStateConfig"
+)
+public class TbDeviceStateNode implements TbNode {
+
+    private static final Set<TbMsgType> SUPPORTED_EVENTS = EnumSet.of(
+            TbMsgType.CONNECT_EVENT, TbMsgType.ACTIVITY_EVENT, TbMsgType.DISCONNECT_EVENT, TbMsgType.INACTIVITY_EVENT
+    );
+    private static final Duration ONE_SECOND = Duration.ofSeconds(1L);
+    private static final Duration ENTRY_EXPIRATION_TIME = Duration.ofDays(1L);
+    private static final Duration ENTRY_CLEANUP_PERIOD = Duration.ofHours(1L);
+
+    private Stopwatch stopwatch;
+    private ConcurrentMap<DeviceId, Duration> lastActivityEventTimestamps;
+    private TbMsgType event;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        TbMsgType event = TbNodeUtils.convert(configuration, TbDeviceStateNodeConfiguration.class).getEvent();
+        if (event == null) {
+            throw new TbNodeException("Event cannot be null!", true);
+        }
+        if (!SUPPORTED_EVENTS.contains(event)) {
+            throw new TbNodeException("Unsupported event: " + event, true);
+        }
+        this.event = event;
+        lastActivityEventTimestamps = new ConcurrentHashMap<>();
+        stopwatch = Stopwatch.createStarted();
+        scheduleCleanupMsg(ctx);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) {
+        if (msg.isTypeOf(TbMsgType.DEVICE_STATE_STALE_ENTRIES_CLEANUP_SELF_MSG)) {
+            removeStaleEntries();
+            scheduleCleanupMsg(ctx);
+            return;
+        }
+
+        if (!EntityType.DEVICE.equals(msg.getOriginator().getEntityType())) {
+            ctx.tellSuccess(msg);
+            return;
+        }
+
+        DeviceId originator = new DeviceId(msg.getOriginator().getId());
+
+        lastActivityEventTimestamps.compute(originator, (__, lastEventTs) -> {
+            Duration now = stopwatch.elapsed();
+
+            if (lastEventTs == null) {
+                sendEventAndTell(ctx, originator, msg);
+                return now;
+            }
+
+            Duration elapsedSinceLastEventSent = now.minus(lastEventTs);
+            if (elapsedSinceLastEventSent.compareTo(ONE_SECOND) < 0) {
+                ctx.tellSuccess(msg);
```

Comment:
rateLimit + add to annotation the corresponding relation types

Reply:
Please see https://github.com/thingsboard/thingsboard/pull/9030/commits/e55d14325a7658193eca00115c7f918098a8406c